File: itw2027/fixed_task/test_noisy_nt60.py
import json
import tempfile
import unittest
from pathlib import Path
from noisy_nt60 import generate


class Noisy60Test(unittest.TestCase):
    def test_fresh_banks_all_frames_and_stable_snr_indices(self):
        with tempfile.TemporaryDirectory() as directory:
            base=Path(directory)/'base'; base.mkdir()
            grids={'bfc':[-4.5,2.], 'conventional':[-4.5,2.]}
            (base/'design.json').write_text(json.dumps(dict(grids=grids)))
            old=[dict(family=f,scheme=s,snr_db=x,snr_index=i+7)
                 for f in ('id','rank','exact')
                 for s in (('bfc','conventional') if f=='exact' else ('bfc',))
                 for i,x in enumerate(grids[s])]
            (base/'noisy.json').write_text(json.dumps(old))
            out=Path(directory)/'new_n_t_60'; generate(base,out)
            banks=json.loads((out/'banks_n_t_60.json').read_text())
            tasks=json.loads((out/'noisy_n_t_60.json').read_text())
            self.assertEqual(len(banks),12); self.assertEqual(len(tasks),32)
            self.assertFalse((out/'noiseless.json').exists())
            for task in tasks:
                self.assertEqual((task['nt'],task['G'],task['frames']),(60,360,2500))
                self.assertTrue(task['output'].endswith('_n_t_60.mat'))
                self.assertIn(task['bank'],[b['output'] for b in banks])
                self.assertEqual(task['snr_index'],7 if task['snr_db']==-4.5 else 8)
                self.assertEqual(task['ldpc_rate'],1/3 if task['scheme']=='bfc' else 3/5)
            self.assertEqual({t['first_frame'] for t in tasks},{1,2501,5001,7501})
            with self.assertRaises(ValueError): generate(base,out)

    def test_corrected_conventional_only(self):
        from scipy.io import savemat
        from conventional_nt60_corrected import generate as corrected
        with tempfile.TemporaryDirectory() as directory:
            base=Path(directory)/'base'; base.mkdir()
            (base/'noisy_n_t_60.json').write_text(json.dumps([
                dict(scheme='conventional',snr_db=-5.5,snr_index=1),
                dict(scheme='conventional',snr_db=0.,snr_index=2)]))
            for shard in range(4):
                p=base/f'banks_n_t_60/exact_{shard:03d}_n_t_60.mat'; p.parent.mkdir(exist_ok=True)
                savemat(p,{'result':dict(complete=True,frames_done=2500,
                    config=dict(nt=60,G=360,m=100),task=dict(seed_group=3,first_frame=2500*shard+1))})
            out=Path(directory)/'corrected'; corrected(base,out)
            tasks=json.loads((out/'noisy_conventional_rc_3_5_n_t_60.json').read_text())
            self.assertTrue(all(t['ldpc_rate']==3/5 and t['scheme']=='conventional' for t in tasks))
            pilot=json.loads((out/'pilot.json').read_text()); self.assertEqual(len(pilot),5)
            self.assertEqual(sum(t['frames'] for t in pilot),12500)


if __name__=='__main__': unittest.main()

File: itw2027/fixed_task/test_noisy_nt60.m
function test_noisy_nt60()
ft_setup(); out=tempname; mkdir(out); cleanup=onCleanup(@() rmdir(out,'s'));
for family={'id','rank','exact'}
    task=struct('kind','bank','family',family{1},'nt',60,'G',360, ...
        'frames',2,'first_frame',1,'seed_group',3,'runtime_limit',600, ...
        'output',[family{1} '_n_t_60.mat']);
    bank=ft_bank(task,out); d=bank.config;
    assert(d.r==30 && d.T==2^30 && d.G==360 && d.neff==180 && d.Ni==21600);
    assert(d.m+d.pad==d.K*d.r);
    assert(size(bank.u,1)==360 && sum(bank.positive)==180);
    % Ensure uint32 packing preserves the entire 30-bit index/tag range.
    u=uint32([1;2^30]); c=uint32([0;2^30-1]);
    payload=pack_bfc_tuples(u,c,30,120,2); [ur,cr]=unpack_bfc_tuples(payload,30);
    assert(isequal(u,ur) && isequal(c,cr));
    for snr=[-8 15]
        noisy=struct('kind','noisy','family',family{1},'scheme','bfc','nt',60,'G',360, ...
            'frames',2,'first_frame',1,'seed_group',3,'snr_db',snr,'snr_index',1, ...
            'bank',task.output,'output',sprintf('%s_%d_n_t_60.mat',family{1},snr),'runtime_limit',600);
        result=ft_noisy(noisy,out);
        assert(result.counts.positive_trials==360 && result.counts.negative_trials==360);
        if snr==15
            assert(result.metrics.FN==0 && result.metrics.FER==0);
            assert(result.counts.false_positives==result.counts.noiseless_false_positives);
        else
            assert(result.metrics.FER==1 && result.counts.false_negatives>0);
        end
    end
    if strcmp(family{1},'exact')
        noisy.scheme='conventional'; noisy.output='conventional_n_t_60.mat';
        noisy.ldpc_rate=3/5;
        result=ft_noisy(noisy,out);
        assert(result.channel.Ni==38880 && result.channel.Rc==3/5);
        assert(result.channel.padding_bits==2880 && result.channel.payload_bits==36000);
        assert(result.metrics.FER==0 && result.metrics.FN==0 && result.metrics.FP==0);
        noisy.snr_db=-8; noisy.output='conventional_low_n_t_60.mat';
        result=ft_noisy(noisy,out); assert(result.metrics.FER==1 && result.metrics.FN>0);
    end
end
legacy=ft_config('id'); assert(legacy.G==540); % Old default remains unchanged.
fprintf('Noisy-only nt=60 tests passed.\n');
end
